{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "285945e6-88d6-492c-8e7b-229d3e76f4e7",
   "metadata": {},
   "source": [
    "#  Preparation of T-cell Enriched annData Object\n",
    "\n",
    "- Cellranger multi (10x Genomics, v7.0.0) and Cellranger aggregate (10x Genomics, v7.0.0) were used on GEX and VDJ libraries from 5 lung samples and 6 liver samples. \n",
    "- The Cellranger multi pipeline was run on the Terra Cloud Computing Platform can be found here: https://portal.firecloud.org/?return=terra#methods/Alignment/Cellranger_Multi/11.\n",
    "- The Cellranger aggregate pipeline was run on the Terra Cloud Computing Platformcan be found here: https://portal.firecloud.org/?return=terra#methods/Alignment/Cellranger_Aggregate_For_Multi_Samples/16\n",
    "- During QC, one liver sample was thrown out due to poor quality (gem well 5). Thus, 5 lung samples and 5 liver samples were included in the analysis\n",
    "- To remove ambient RNA, we then ran the \"remove-background\" command in Cellbender. Default settings were used, with fpr = \"0.01\" and total_droplets_included = 40000. The pipeline was run on the Terra Cloud Computing platform and can be found here: https://portal.firecloud.org/?return=terra#methods/cellbender/remove-background/11\n",
    "- The resulting data was incorporated into a annData object (v0.9.2)\n",
    "- During manual QC, only Cell-IDs with mitochondrial DNA < 10%, a minimum of 750 genes, and at least 2000 transcripts but no greater than 25000 transcripts were retained\n",
    "- Using pybiomart (version 0.2.0), we extracted Ensembl Biomart was used to identify Ensembl gene IDs, gene names, gene biotypes, and chromosome names. Pybiomart dataset settings were as follows: name = 'mmulatta_gene_ensembl, host='http://www.ensembl.org'\n",
    "- Only genes that were \"protein_coding\" and \"miRNA\" gene types were retained. Genes from the Y chromosome were removed.\n",
    "- Scanpy (insert version) was used to identify highly variable genes from raw counts, with \"n_top_genes = 5000\", \"flavor = seurat_v3\", and \"subset=False\" \n",
    "- Only highly variable genes as well as the following T-cell markers were carried forward: \"CD3E\", \"TRAC\", \"CD4\", and \"CD8A\".\n",
    "- The .hdf5 file containing this filtered object has been provided as **\"01_pre_filtered_object.hdf5\"**. It has been modified for usability.\n",
    "- A VAE model was trained using the pre_filtered_object via scVI. \n",
    "- There were two experimental batches. Categorical covariate keys were experimental batch as well as animal.\n",
    "- Non T-cell enriched clusters were thrown out and a new VAE model was trained, followed by another round of scVI clustering. The final filtered object is provided as **\"02_final_filtered_object.hdf5\"**. It has been modified for usability. It is this object that was used for all downstream analysis.\n",
    "- The two VAE models are provided as **\"scVI_Model_01\"** and **\"scVI_Model_02\"**. Please note that the models were initially trained on <v0.15.0. To use the models in a newer version, the \"convert_legacy_save\" function must be run to convert to an updated format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97744e42-06fa-4c94-a874-7739f93d1ce6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np #v1.24.4\n",
    "import pandas as pd #v2.1.1\n",
    "import anndata #v0.9.2\n",
    "import scanpy as sc #v1.9.6\n",
    "import scvi #v<0.15.0\n",
    "import random\n",
    "\n",
    "c_iSeed = 6161904\n",
    "np.random.seed(c_iSeed)\n",
    "random.seed(c_iSeed)\n",
    "scvi.settings.num_threads = 8\n",
    "scvi.settings.progress_bar_style = \"rich\"\n",
    "scvi.settings.seed = c_iSeed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a60118a4-bb1b-49a6-ac15-c8240bfd231d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pre_filtered_object = anndata.read_h5ad(filename=___) #Replace ___ with path to file \"01_pre_filtered_object.hdf5\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b54c49c-2c8b-495e-b367-31ec7d9679d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set scVI Model Parameters\n",
    "scvi.model.SCVI.setup_anndata(pre_filtered_object,\n",
    "    layer=\"raw_counts\",\n",
    "    categorical_covariate_keys=[\"experimental_batch\", \"animal_id\"]\n",
    ")\n",
    "\n",
    "scVI_Model_01 = scvi.model.SCVI(pre_filtered_object,\n",
    "                          n_layers=2,\n",
    "                          n_latent=30,\n",
    "                          gene_likelihood=\"nb\")\n",
    "\n",
    "n_val_max = 40000\n",
    "if 0.1 * pre_filtered_object.n_obs < n_val_max:\n",
    "    train_size = 0.9\n",
    "else:\n",
    "    train_size = 1 - (n_val_max/pre_filtered_object.n_obs)\n",
    "\n",
    "# Train VAE\n",
    "scVI_Model_01.train(train_size=train_size,\n",
    "              check_val_every_n_epoch=10,\n",
    "              early_stopping=True,    \n",
    "              early_stopping_patience=45,\n",
    "              max_epochs=10000, \n",
    "              batch_size=1024,    \n",
    "              limit_train_batches=20)\n",
    "\n",
    "# Generate scVI Clusters\n",
    "def Prepare_scVI_Latent_Space(annIn,vaeIn):\n",
    "\n",
    "    # Get Latent Space\n",
    "    annIn.obsm[\"scVI_Latent_Space\"] = vaeIn.get_latent_representation()\n",
    "            \n",
    "    # Cluster\n",
    "    sc.pp.neighbors(annIn, use_rep=\"scVI_Latent_Space\")\n",
    "    sc.tl.umap(annIn) #For visualization downstream\n",
    "    sc.tl.leiden(annIn, key_added=\"scVI_clusters\")\n",
    "    \n",
    "    return\n",
    "\n",
    "Prepare_scVI_Latent_Space(pre_filtered_object,scVI_Model_01)\n",
    "\n",
    "# Get Normalized Expression\n",
    "pre_filtered_object.layers[\"scvi_norm_counts\"] = scVI_Model_01.get_normalized_expression(library_size=10e4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24ebef14-5e24-474f-82e6-9386df3b90c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Throw Out Non-T Cell Enriched Clusters \n",
    "pre_filtered_object = pre_filtered_object[pre_filtered_object.obs[\"scVI_clusters\"] != 16]\n",
    "pre_filtered_object = pre_filtered_object[pre_filtered_object.obs.query(\"scVI_clusters != '16'\").index]\n",
    "pre_filtered_object = pre_filtered_object[[r != \"16\" for r in pre_filtered_object.obs.scVI_clusters]]\n",
    "\n",
    "pre_filtered_object = pre_filtered_object[pre_filtered_object.obs[\"scVI_clusters\"] != 17]\n",
    "pre_filtered_object = pre_filtered_object[pre_filtered_object.obs.query(\"scVI_clusters != '17'\").index]\n",
    "pre_filtered_object = pre_filtered_object[[r != \"17\" for r in pre_filtered_object.obs.scVI_clusters]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78a9be8c-90a6-47d3-9317-eff0654b1bca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set scVI Model Parameters\n",
    "scvi.model.SCVI.setup_anndata(pre_filtered_object,\n",
    "    layer=\"raw_counts\",\n",
    "    categorical_covariate_keys=[\"experimental_batch\", \"animal_id\"]\n",
    ")\n",
    "\n",
    "scVI_Model_02 = scvi.model.SCVI(pre_filtered_object,\n",
    "                          n_layers=2,\n",
    "                          n_latent=30,\n",
    "                          gene_likelihood=\"nb\")\n",
    "\n",
    "n_val_max = 40000\n",
    "if 0.1 * post_scVI_object.n_obs < n_val_max:\n",
    "    train_size = 0.9\n",
    "else:\n",
    "    train_size = 1 - (n_val_max/pre_filtered_object.n_obs)\n",
    "\n",
    "# Train VAE\n",
    "scVI_Model_02.train(train_size=train_size,\n",
    "              check_val_every_n_epoch=10,\n",
    "              early_stopping=True,    \n",
    "              early_stopping_patience=45,\n",
    "              max_epochs=10000, \n",
    "              batch_size=1024,    \n",
    "              limit_train_batches=20)\n",
    "\n",
    "# Generate scVI Clusters\n",
    "def Prepare_scVI_Latent_Space(annIn,vaeIn):\n",
    "\n",
    "    # Get Latent Space\n",
    "    annIn.obsm[\"scVI_Latent_Space\"] = vaeIn.get_latent_representation()\n",
    "            \n",
    "    # Cluster\n",
    "    sc.pp.neighbors(annIn, use_rep=\"scVI_Latent_Space\")\n",
    "    sc.tl.umap(annIn) #For visualization downstream\n",
    "    sc.tl.leiden(annIn, key_added=\"scVI_clusters\")\n",
    "    \n",
    "    return\n",
    "\n",
    "Prepare_scVI_Latent_Space(pre_filtered_object,scVI_Model_02)\n",
    "\n",
    "# Get Normalized Expression\n",
    "pre_filtered_object.layers[\"scvi_norm_counts\"] = scVI_Model_02.get_normalized_expression(library_size=10e4)\n",
    "\n",
    "# The final annData object is provided as \"02_final_filtered_object.hdf5\" and was used for all downstream analysis"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
